File: snippets/python/src/services/configuration.py
from datetime import timedelta

import restate

# <start_options>
# Specify service options when you create them
my_workflow = restate.Workflow(
    "MyWorkflow",
    inactivity_timeout=timedelta(minutes=15),
    abort_timeout=timedelta(minutes=15),
    idempotency_retention=timedelta(days=3),
    journal_retention=timedelta(days=7),
    ingress_private=True,
    enable_lazy_state=True,  # only for Objects/Workflows
    invocation_retry_policy=restate.InvocationRetryPolicy(
        initial_interval=timedelta(seconds=1),
        max_interval=timedelta(seconds=30),
        max_attempts=10,
        on_max_attempts="pause"
    )
)
# <end_options>


# <start_handleropts>
# Specify handler options via the decorator
@my_workflow.main(
    inactivity_timeout=timedelta(minutes=15),
    abort_timeout=timedelta(minutes=15),
    workflow_retention=timedelta(days=3),
    # -> or idempotency_retention for Services/Objects
    journal_retention=timedelta(days=7),
    ingress_private=True,
    enable_lazy_state=True,  # only for Objects/Workflows
    invocation_retry_policy=restate.InvocationRetryPolicy(
        initial_interval=timedelta(seconds=1),
        max_interval=timedelta(seconds=30),
        max_attempts=10,
        on_max_attempts="pause"
    )
)
async def run(ctx: restate.WorkflowContext, req: str) -> str:
    # ... implement workflow logic here ---
    return "success"


#  <end_handleropts>


app = restate.app([my_workflow])

if __name__ == "__main__":
    import hypercorn
    import hypercorn.asyncio
    import asyncio

    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))

File: snippets/go/services/configuration.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
)

type MyWorkflow struct {
}

func (MyWorkflow) MyHandler(ctx restate.WorkflowContext, greeting string) (string, error) {
	return fmt.Sprintf("%s!", greeting), nil
}

func main() {
	// <start_options>
	// Specify service options when binding them to an endpoint
	if err := server.NewRestate().
		Bind(
			restate.Reflect(
				MyWorkflow{},
				restate.WithInvocationRetryPolicy(
					restate.WithInitialInterval(time.Second),
					restate.WithMaxInterval(30*time.Second),
					restate.WithMaxAttempts(10),
					restate.PauseOnMaxAttempts()),
				restate.WithInactivityTimeout(15*time.Minute),
				restate.WithAbortTimeout(15*time.Minute),
				restate.WithIdempotencyRetention(3*24*time.Hour),
				restate.WithJournalRetention(7*24*time.Hour),
				restate.WithIngressPrivate(true),
				restate.WithEnableLazyState(true),
				restate.WithWorkflowRetention(10*24*time.Hour), // Only for workflows
			),
		).
		Start(context.Background(), "0.0.0.0:9080"); err != nil {
		log.Fatal(err)
	}
	// <end_options>

	// <start_handleropts>
	// Use ConfigureHandler to customize a handler configuration
	if err := server.NewRestate().
		Bind(
			restate.Reflect(MyWorkflow{}).
				ConfigureHandler("run",
					restate.WithInvocationRetryPolicy(
						restate.WithInitialInterval(time.Second),
						restate.WithMaxInterval(30*time.Second),
						restate.WithMaxAttempts(10),
						restate.PauseOnMaxAttempts()),
					restate.WithInactivityTimeout(15*time.Minute),
					restate.WithAbortTimeout(15*time.Minute),
					restate.WithIdempotencyRetention(3*24*time.Hour),
					restate.WithJournalRetention(7*24*time.Hour),
					restate.WithIngressPrivate(true),
					restate.WithEnableLazyState(true),
					restate.WithWorkflowRetention(10*24*time.Hour), // Only for workflows
				),
		).
		Start(context.Background(), "0.0.0.0:9080"); err != nil {
		log.Fatal(err)
	}
	// <end_handleropts>
}
